### PR TITLE
Add support for defining status as a subresource of flyteworkflow CRD

### PR DESCRIPTION
**Pull request by hamersaw**
# TL;DR
Adding support for setting status as a subresource on the flyteworkflow CRD. This should reduce traffic to the kubernetes API server by reducing update message sizes.

## Type
 - [X] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [X] Smoke tested
 - [X] Unit tests added
 - [X] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Setting status as a subresource on the flyteworkflow CRD introduces a separate kubernetes rest API which partitions updates of the CRD spec and status subfields. This functionality seems straight-forward, but introduces complexities when trying to provide backward compatible support for existing deployments where the CRD doesn't have the status defined as a subresource. Specifically, the cases are described below for both the existing and non-existing CRD status subresource and the effect of the Update and UpdateStatus API calls.

case 1: subresource exists on CRD
    1.1 client.Update(...) -> succeeds, but status does not change - update doesn't care about changing status subresource
    1.2 client.UpdateStatus(...) -> succeeeds
case 2: no subresouce on CRD
    2.1 -> client.Update(...) -> succeeds (status does change)
    2.2 -> client.UpdateStatus(...) -> throws k8s error NotFound  - impossible to distinguish from 1.2 where the resource does not actually exist

Therefore, we determine if the status subresource API is available and use it to inform the correct commands to update flyteworkflow CRD status and spec subfields. This means the implementation works for the existing flyteworkflow CRD definition (ie. without status subresource) and offers advantages once the subresource is added.

## Tracking Issue
https://github.com/flyteorg/flyte/issues/126

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by hamersaw** [comment]
@EngHabu is it sane to structure the hasStatusSubresourceAPI handling in the controller logic? i'm trying to convince myself this approach is better than burying it in the workflowstore Update function. is there a compelling reason why workflowstore has separate Update and UpdateStatus functions besides mimicing the kubernetes API? logically, the controller shouldn't care what API calls update the flyteworkflow CRD, just that it happens correctly.

**Comment by hamersaw** [comment]
It was decided that at the moment support for this feature faces complications in deployment and upgrading of flytepropeller and flyte kubernetes components. See https://github.com/flyteorg/flyte/issues/126 for further explanation.
